src/v3/src/util/theme.ts

```diff
@@ -10,116 +10,429 @@
-          .set('hsl.s', '+0.18')
-          .set('hsl.l', '-0.24')
-          .hex(),
-        inverseTextColor: getInverseTextColor(brand.primaryColor),
-      };
-    }
-
-    return {
-      primaryColor: brand.primaryColor,
-      textColor: '#1d1d21',
-      primaryColorLight: chroma(brand.primaryColor)
-        .set('hsl.l', '+0.62')
+        .set('hsl.l', lightness < 0.59 ? '+0.31' : 0.9) // clamp lightness
+        .hex(),
+      dark: chroma(main)
+        .set('hsl.h', '+3')
+        .set('hsl.s', '+0.18')
+        .set('hsl.l', '-0.24')
         .hex(),
-      primaryColorLightest: chroma(brand.primaryColor)
+      contrastText: getInverseTextColor(main),
+    } : {
+      main: main,
+      lighter: chroma(main)
         .set('hsl.h', '+9')
         .set('hsl.s', '+0.18')
         .set('hsl.l', 0.97)
         .hex(),
-      primaryColorDark: chroma(brand.primaryColor)
+      light: chroma(main)
+        .set('hsl.l', '+0.62')
+        .hex(),
+      dark: chroma(main)
         .set('hsl.l', '+0.31')
         .hex(),
-      inverseTextColor: getInverseTextColor(brand.primaryColor),
-    };
-  } catch (e) {
-    // eslint-disable-next-line no-console
-    console.warn('Invalid theme color in configuration', brand);
-
-    return null;
-  }
+      contrastText: getInverseTextColor(main),
+    }
 };
 
-export const mapMuiThemeFromBrand = (
-  brand: BrandColors | undefined,
-  languageDirection: LanguageDirection,
-  muiThemeOverrides?: ThemeOptions,
+export const createTheme = (
+  brand?: Brand,
+  customTokens?: Partial<DesignTokensType>,
 ): Theme => {
-  // TODO: OKTA-517723 temporary override until odyssey-react-mui theme borderRadius value is fixed
-  odysseyTheme.shape.borderRadius = 4;
+  const defaultTokens = DESIGN_TOKENS;
+  const mergedTokens: DesignTokensType = { ...defaultTokens, ...customTokens };
+  const theme = odysseyTheme;
+  theme.palette.text.primary = BLACK;
+  if (brand?.primaryColor) {
+    const palettePrimary = generatePalette(brand.primaryColor);
+    const { lighter, light, main, dark } = palettePrimary;
+    set(theme, 'palette.primary.lighter', lighter);
+    set(theme, 'palette.primary.light', light);
+    set(theme, 'palette.primary.main', main);
+    set(theme, 'palette.primary.dark', dark);
+  }
+  if (customTokens) {
+    if (customTokens.PalettePrimaryMain) {
+      const { lighter, light, main, dark } = generatePalette(customTokens.PalettePrimaryMain);
+      set(theme, 'palette.primary.lighter', customTokens.PalettePrimaryLighter ?? lighter);
+      set(theme, 'palette.primary.light', customTokens.PalettePrimaryLight ?? light);
+      set(theme, 'palette.primary.main', customTokens.PalettePrimaryMain ?? main);
+      set(theme, 'palette.primary.dark', customTokens.PalettePrimaryDark ?? dark);
+      mergedTokens.PalettePrimaryLighter = lighter;
+      mergedTokens.PalettePrimaryLight = light;
+      mergedTokens.PalettePrimaryMain = main;
+      mergedTokens.PalettePrimaryDark = dark;
```

Comment:
Writing generated palette colors back to design tokens so they're picked up by `MuiButton` themeOverrides below